### PR TITLE
Add maxwell_dominant_planewave

### LINKED ISSUE #69: provide dominant planewave component
For various calculations, it would be convenient to have a function to return the dominant planewave component.   That is, find the **G** vector corresponding to the biggest amplitude planewave (in the planewave expansion of an eigenmode), and return **k**+**G** (the corresponding planewave wavevector).

Has to be implemented at the C level, as a new function in `src/maxwell`, but should be straightforward since we already have all the requisite information.

(Would help with stevengj/meep#476, for example.)

### PR DESCRIPTION
**Pull request by ChristopherHogan**
Closes #69. After #70 is merged, I can rebase this and build it in MPI mode on travis. Is there a simple test I can add, or do we have to wait for higher level features?
@stevengj @oskooi

### CODE REVIEW

**Comment by stevengj** [comment]
Should be possible to test in https://github.com/stevengj/mpb/blob/master/tests/maxwell_test.c … the dominant wavevector in this simple 1d case should be (k,0,0), where k is (k,k-1,k+1,k-2,k+2,…) for bands (1,2,3,…).

**Comment by ChristopherHogan** [comment]
I added the test you suggested, but I'm not getting the expected results:
```
Expected kdom:        0.500000       0.000000       0.000000
Got kdom:            -0.500000       0.000000       0.000000
Expected kdom:        1.500000       0.000000       0.000000
Got kdom:            -0.500000       0.000000       0.000000
Expected kdom:       -0.500000       0.000000       0.000000
Got kdom:            -0.500000       0.000000       0.000000
Expected kdom:        2.500000       0.000000       0.000000
Got kdom:            -0.500000       0.000000       0.000000
Expected kdom:       -1.500000       0.000000       0.000000
Got kdom:            -1.500000       0.000000       0.000000
Expected kdom:        3.500000       0.000000       0.000000
Got kdom:             1.500000      -0.000000       0.000000
Expected kdom:       -2.500000       0.000000       0.000000
Got kdom:            -1.500000       0.000000       0.000000
Expected kdom:        4.500000       0.000000       0.000000
Got kdom:            -1.500000       0.000000       0.000000
```
I'm not sure if it's an issue with the test or the implementation.

**Comment by stevengj** [comment]
kx=0.5 is a special case.  Try kx=0.4

**Comment by ChristopherHogan** [comment]
Changing `maxwell_test.c:55` from `#define KX 0.5` to `#define KX 0.4` results in 
```
Expected kdom:        0.400000       0.000000       0.000000
Got kdom:             0.400000      -0.000000       0.000000
Expected kdom:        1.400000       0.000000       0.000000
Got kdom:             0.400000      -0.000000       0.000000
Expected kdom:       -0.600000       0.000000       0.000000
Got kdom:            -0.600000       0.000000       0.000000
Expected kdom:        2.400000       0.000000       0.000000
Got kdom:            -0.600000       0.000000       0.000000
Expected kdom:       -1.600000       0.000000       0.000000
Got kdom:             1.400000      -0.000000       0.000000
Expected kdom:        3.400000       0.000000       0.000000
Got kdom:             1.400000      -0.000000       0.000000
Expected kdom:       -2.600000       0.000000       0.000000
Got kdom:            -1.600000       0.000000       0.000000
Expected kdom:        4.400000       0.000000       0.000000
Got kdom:            -1.600000       0.000000       0.000000
```

**Comment by stevengj** [comment]
@ChristopherHogan, the bands are doubled because the default is NO_PARITY, so you get two polarizations per band.  Pass the `-e` option.

I'm no sure why you are printing 1.4 as the second "expected" kdom — note in my message above that it should be (kx, kx-1, kx+1, …), which gives (0.4, -0.6, 1.4, …).  So, it looks like it is giving the correct results.

**Comment by stevengj** [comment]
Note that you can change kx to 0.4 by passing `-k 0.4`, without editing `KX` in the source code.

**Comment by ChristopherHogan** [comment]
I made the fixes you suggested, and now the first 4 bands are correct, but the last 4 are not.
```
$ ./maxwell_test -1 -c 1e-9 -x 256 -E 1e-3 -e -k 0.4
...
Expected kdom:        0.400000       0.000000       0.000000
Got kdom:             0.400000      -0.000000       0.000000
Expected kdom:       -0.600000       0.000000       0.000000
Got kdom:            -0.600000       0.000000       0.000000
Expected kdom:        1.400000       0.000000       0.000000
Got kdom:             1.400000      -0.000000       0.000000
Expected kdom:       -1.600000       0.000000       0.000000
Got kdom:            -1.600000       0.000000       0.000000
Expected kdom:        2.400000       0.000000       0.000000
Got kdom:             1.400000      -0.000000       0.000000
Expected kdom:       -2.600000       0.000000       0.000000
Got kdom:            -1.600000       0.000000       0.000000
Expected kdom:        3.400000       0.000000       0.000000
Got kdom:             2.400000      -0.000000       0.000000
Expected kdom:       -3.600000       0.000000       0.000000
Got kdom:            -2.600000       0.000000       0.000000
```

File: src/maxwell/maxwell.c [hunk 1]
```diff
@@ -308,6 +308,48 @@ static void compute_cross(real *a0, real *a1, real *a2,
      *a2 = b0 * c1 - b1 * c0;
 }
 
+typedef struct {
+    double amp;
+    int rank;
+} amp_and_rank;
+
+void maxwell_dominant_planewave(maxwell_data *d, evectmatrix H, int band, double kdom[3])
```
Comment (stevengj) [inline src/maxwell/maxwell.c]: add a comment that kdom is in Cartesian coordinates

**Comment by ChristopherHogan** [comment]
Running separate maxwell_test for maxwell_dominant_planewave. This is green and ready.

File: src/maxwell/maxwell.c (hunk 1 above)
Comment (ChristopherHogan) [reply src/maxwell/maxwell.c]: Fixed.

**Pull request merged**